### PR TITLE
Ensure initial binding always binds to natural type

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1198816/, which is about `CSharpOperationFactory.CreateInternal` being passed a node it does not expect. I looked through all the places in the binder where we use `BindValue` and ensured the results were being correctly bound to a natural type, even if there was some error. I found 2 such cases:

- Array rank specifiers were not being bound to their natural type if they had some error.
- `if` conditions were not being bound to their natural type if they had some error. This is the case that results in the specific stack pattern from the watson, where a `BoundConversion` sat on top of a `BoundUnconvertedObjectCreationExpression`.

In both of these cases, unconverted results were passed to the factory, which cannot handle nodes of these types and threw.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for reviews please.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConditionalOperatorTests.cs [hunk 1]
```diff
@@ -1499,5 +1499,24 @@ public TestClass Self()
...
+        _ = new(bad) ? null : new object();
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (5,17): error CS0103: The name 'bad' does not exist in the current context
+                //         _ = new(bad) ? null : new object();
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "bad").WithArguments("bad").WithLocation(5, 17)
+            );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConditionalOperatorTests.cs:1520]: >} [](start = 8, length = 1)

We should test other expressions the rely on `BindToNaturalType()`.

For instance:
```C#
class Program
{
    static void Main()
    {
        _ = (new(bad), null) ? null : new object();
    }
}
```

And:
```C#
class Program
{
    static void F(int i)
    {
        _ = i switch { 0 => new(bad), _ => true } ? null : new object();
    }
}
```

**Review by cston — APPROVED** [review]

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please.

**Comment by AlekseyTs** [comment]
@333fred 

> In both of these cases, unconverted results were passed to the factory, which cannot handle nodes of these types and threw.

I think @gafter added an assert somewhere in flow analysis to catch all the nodes that haven't been properly adjusted. I am curious if these cases are caught by that assert. If they are not, perhaps it would be useful to understand why and adjust the assert accordingly.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -3165,6 +3165,10 @@ private BoundExpression BindArrayDimension(ExpressionSyntax dimension, Diagnosti
                         hasErrors = true;
                     }
                 }
+                else
+                {
+                    size = BindToNaturalType(size, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3170]: >BindToNaturalType [](start = 27, length = 17)

BindToTypeForErrorRecovery

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3170]: >diagnostics [](start = 51, length = 11)

I don't think we want to collect diagnostics on this code path.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

**Comment by 333fred** [comment]
> I think gafter added an assert somewhere in flow analysis to catch all the nodes that haven't been properly adjusted. I am curious if these cases are caught by that assert. If they are not, perhaps it would be useful to understand why and adjust the assert accordingly.

Running the added tests without my fix triggers that assert. It looks like we didn't have test coverage for this path. It also looks like the path is missing checking for unconverted address of operators (it was added in Neal's branch, which was in development at the same time as mine, so the address of operator was never added). I'll add that and a test here, as that would also crash.

**Comment by 333fred** [comment]
> It also looks like the path is missing checking for unconverted address of operators (it was added in Neal's branch, which was in development at the same time as mine, so the address of operator was never added). I'll add that and a test here, as that would also crash.

Actually, I'm going to do this change on master. It looks like it's all handled appropriately currently.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 5)

**Pull request merged**